src/main/java/org/folio/uk/service/UserService.java

```diff
@@ -98,21 +106,55 @@ public Optional<User> getUser(UUID id) {
     return usersClient.lookupUserById(id);
   }
 
-  public CompositeUser getUserBySelfReference(List<IncludedField> include, boolean expandPermissions) {
+  public CompositeUser getUserBySelfReference(List<IncludedField> include, boolean expandPermissions,
+                                              boolean overrideUser) {
     log.info("Retrieving user by self reference with parameters: include = {}, expandPermissions = {}",
       () -> StringUtils.join(emptyIfNull(include), ", "), () -> expandPermissions);
 
     var userId = getUserId();
-
     var user = usersClient.lookupUserById(userId)
       .orElseThrow(() -> new EntityNotFoundException("User was Not Found with: id = " + userId));
 
-    return new CompositeUser()
-      .user(user)
+    // Whether to override from shadow to a real user in ECS
+    if (Boolean.TRUE.equals(overrideUser) && isShadowUserType(user)) {
+      var originalTenantIdOptional = user.getCustomFields().entrySet().stream()
+        .filter(entry -> entry.getKey().equalsIgnoreCase(ORIGINAL_TENANT_ID_CUSTOM_FIELD))
+        .map(entry -> (String) entry.getValue()).findFirst();
+      if (originalTenantIdOptional.isPresent()) {
+        return getRealUserByReference(user, originalTenantIdOptional.get(), expandPermissions);
+      }
+    }
+
+    return new CompositeUser().user(user)
       .permissions(fetchPermissionUser(userId, expandPermissions))
       .servicePointsUser(fetchServicePointUser(userId));
   }
 
+  private boolean isShadowUserType(User user) {
+    return StringUtils.isNotEmpty(user.getType()) && user.getType().equalsIgnoreCase(SHADOW_USER_TYPE);
```

Comment:
can be simplified that also avoids NPE: StringUtils.equals(user.getType(), SHADOW_USER_TYPE);